applications/StructuralMechanicsApplication/python_scripts/structural_response.py

```diff
@@ -220,7 +220,215 @@ def GetShapeGradient(self):
         return gradient
 
 # ==============================================================================
+class AdjointResponseFunction(ResponseFunctionBase):
+    """Linear static adjoint response function.
+    - runs the primal analysis (writes the primal results to an .h5 file)
+    - reads the primal results from the .h5 file into the adjoint model part
+    - uses primal results to calculate value
+    - uses primal results to calculate gradient by running the adjoint analysis
 
+    Attributes
+    ----------
+    primal_analysis : Primal analysis object of the response function
+    adjoint_analysis : Adjoint analysis object of the response function
+    response_function_utility: Cpp utilities object doing the actual computation of response value and gradient.
+    """
+    def __init__(self, identifier, project_parameters, model_part = None):
+        self.identifier = identifier
+
+        model = Model()
+        if model_part is not None:
+            model.AddModelPart(model_part)
+
+        # Create the primal solver
+        ProjectParametersPrimal = Parameters( open(project_parameters["primal_settings"].GetString(),'r').read() )
+        self.primal_analysis = structural_mechanics_analysis.StructuralMechanicsAnalysis(model, ProjectParametersPrimal)
+        self.primal_model_part_name = ProjectParametersPrimal["problem_data"]["model_part_name"].GetString()
+
+        # Create the adjoint solver
+        ProjectParametersAdjoint = Parameters( open(project_parameters["adjoint_settings"].GetString(),'r').read() )
+        ProjectParametersAdjoint["solver_settings"].AddValue("response_function_settings", project_parameters)
+
+        adjoint_model = Model()
+        # TODO find out why it is not possible to use the same model_part
+        self.adjoint_analysis = structural_mechanics_analysis.StructuralMechanicsAnalysis(adjoint_model, ProjectParametersAdjoint)
+        self.adjoint_model_part_name = ProjectParametersAdjoint["problem_data"]["model_part_name"].GetString()
+
+
+    def Initialize(self):
+        self.primal_analysis.Initialize()
+        self.adjoint_analysis.Initialize()
+
+        self.primal_model_part = self.primal_analysis.model.GetModelPart(self.primal_model_part_name)
+        self.adjoint_model_part = self.adjoint_analysis.model.GetModelPart(self.adjoint_model_part_name)
+
+        # TODO should be created here, not in solver!
+        self.response_function_utility = self.adjoint_analysis._GetSolver().response_function
```

Comment:
We need to clarify who is the owner of the response function. The owner should be responsible for calling the response_function.InitializeSolutionStep etc. methods.

Reply:
I improved it, should be more clear now

Reply:
We may want to move ownership to the stage, but I agree with leaving the code as it is now for this pr.

Reply:
`response_function_utility` is now owned by the adjoint solver (in all cases)